### DISCUSSION TITLE
Sharing tips for future3 on bookworm/kernel 6.6 and upstreaming fixes

### DISCUSSION THREAD

**Discussion by hoffie** [MiczFlor/RPi-Jukebox-RFID#2295] [Ideas]
First of all: Thanks a lot to all creators and contributors for developing such a cool project!

I've recently started building a Phoniebox using the future3/main branch on a Raspberry Pi Model 3B with Raspberry Pi OS Bookworm Lite - 32bit, latest kernel (6.6).

I'd like to share some experiences to help others avoid spending as many hours as I did. ;)
- :+1: future3 works reasonably well (I hadn't used v2 before though)
- future3 still has some rough edges or missing features (as documented elsewhere). Most-relevant for me:
  - https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/1946#issuecomment-1972101819
  - https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2294
- :warning: Bookworm ships NetworkManager by default and NetworkManager is rather opinionated when it comes to Wifi configuration. When enabling WPA, NetworkManager also force-enables WPA-SHA256-PSK (and FT-PSK). This breaks Wifi connectivity in some setups (apparently at least in WPA2/WPA3 mixed mode on my OpenWRT-based router; `wlan0: CTRL-EVENT-ASSOC-REJECT bssid=00:00:00:00:00:00 status_code=16`). Configuring `wpa_supplicant` manually is a workaround. I chose to implement a small log watcher to detect such connection problems and fix the config on the fly: https://github.com/hoffie/raspberry-scripts/tree/main/wpa-no-psk-sha256
- :warning: Bookworm (at least with the latest kernel, 6.6) **drops support for Raspberry-specific /sys/class/gpio numbering, which was mandatory for RPi.GPIO** (other distributions seem to have dropped /sys/class/gpio completely by now).
  Background: https://waldorf.waveform.org.uk/2022/the-one-where-dave-breaks-stuff.html
  This breaks several pieces of a typical Phoniebox. A common error message is `RuntimeError: Failed to add edge detection` when trying to listen for events on GPIOs (e.g. when wiring up buttons or using IRQ pins).
  - gpio.yaml is broken by default. While it uses gpiozero, it is configured to use the now broken RPi.GPIO. This is easily fixable by setting a different PIN factor in the config. I chose lgpio.
  - OnOffShim's clean-shutdown doesn't work any longer: https://github.com/pimoroni/clean-shutdown/issues/37
    Luckily, there are nice workarounds:
    `dtoverlay=gpio-poweroff,gpiopin=4,active_low=1` in `/boot/firmware/config.txt` makes the Raspberry cut power after shutdown without any scripts. The shutdown button can be configured via this project's `gpio.yaml`.
  - The python module for the RC522 RFID reader ([pi-rc522](https://github.com/ondryaso/pi-rc522)) won't work out of the box. It might still work when not using `wait_for_tag()` (which is required for `play_not_swipe`). I see two ways forward:
    - If phasing out RPi.GPIO is not feasible, there is a shim which implements the same API as the original RPi.GPIO, but which uses lgpio under the hood and is therefore able to work on recent kernels. See https://rpi-lgpio.readthedocs.io/en/release-0.4/. I haven't fully tested it and I guess it's rather a band-aid solution than a proper way forward.
    - I've built a modified version of pi-rc522 which uses gpiozero as its backend: https://github.com/hoffie/pi-rc522-gpiozero
      While at it, I also reworked the code to use named registers instead of hardcoded addresses and to fix an issue in the IRQ handling which caused to IRQ to continuously fire after the first card read. This is what I currently use in my box.

So, some of these findings are related to Debian, some are related to RC522 and some are related to this very project. Let me know if some of those should be documented somewhere and/or if I should create further Issues/PRs.

[ ] Upstream report Debian/NetworkManager

**Comment by s-martin** [comment]
Thanks for this comprehensive documentation of bookworm issues.

@AlvinSchiller the NetworkManager stuff might be especially interesting as well as GPIO.

@pabera 

I think we should convert this discussion into an issue (should be possible by a GitHub function) or break it into several topics (the only time I miss epics in GitHub ;)

**Reply by AlvinSchiller** [reply]
Fallback for NetworkManager can be to use the "old" dhcpcd. This is changeable in the raspi-config, but needs to be done before the installation currently.

**Reply by hoffie** [reply]
> Fallback for NetworkManager can be to use the "old" dhcpcd. This is changeable in the raspi-config, but needs to be done before the installation currently.

Right. Personally, I was a bit hesistant to use the old-style config as such things will be removed at some point. I think the WPA stuff is something which I should report to Debian and/or NetworkManger upstream. Regarding Phoniebox it should just be mentioned as some "Known issue" somewhere.

Also, it's a bit annoying. You flash your SD card and boot it and unless you've got a screen connected you won't now what is broken. The RPi just won't appear on the network. Putting the SD back into a card reader and reading back the journald-file on a Linux PC is what evantually lead me to discover the actual issue.

**Reply by s-martin** [reply]
If you report it, could you post a link here for reference?

**Reply by AlvinSchiller** [reply]
But this is only a problem with some router?
I have a Pi Zero and a Pi 3A+ and my fritz.box wifi runs also in WPA2/WP3 mixed mode and couldn't observe connection problems.

**Reply by hoffie** [reply]
> But this is only a problem with some router?

Yes, that might well be possible. I did not have a chance to test with multiple routers (at the same time, no other devices on my network have any problems with its configuration).

**Comment by s-martin** [comment]
@AlvinSchiller did you try rc522 with bookworm?

**Reply by AlvinSchiller** [reply]
With the old version, yes. Not yet with the new one. Will look into this

**Reply by AlvinSchiller** [reply]
Updated a Bookworm Installation and RC522 seems to be broken. So can confirm.

This is also a problem for direct usage of RPI.GPIO in V2 for buttons.

**Reply by hoffie** [reply]
> This is also a problem for direct usage of RPI.GPIO in V2 for buttons.

Maybe the mentioned shim would be a good solution as long as v2 is still supported?

**Reply by hoffie** [reply]
Another note: Official Debian bookworm seems to ship kernel 6.1 by default, but a fully updated RPi OS will now default to 6.6. So it's possible that bookworm once worked (while on 6.1), but got broken with 6.6. It seems like the rollout started 2023-12-08: https://forums.raspberrypi.com/viewtopic.php?t=361116 and it became the default (via apt) on 2024-03-11: https://forums.raspberrypi.com/viewtopic.php?t=361116&sid=8673583227775ac65c4eb127cba7de96&start=75#p2202173

**Reply by s-martin** [reply]
Yes, I think that's one option. 

The other option could be that V2 is only supported for Bullseye anymore and we put more efforts in making future3 feature complete.

**Reply by AlvinSchiller** [reply]
> Another note: Official Debian bookworm seems to ship kernel 6.1 by default, but a fully updated RPi OS will now default to 6.6. So it's possible that bookworm once worked (while on 6.1), but got broken with 6.6. It seems like the rollout started 2023-12-08: https://forums.raspberrypi.com/viewtopic.php?t=361116 and it became the default (via apt) on 2024-03-11: https://forums.raspberrypi.com/viewtopic.php?t=361116&sid=8673583227775ac65c4eb127cba7de96&start=75#p2202173

Thats right. It's just a problem sind the new release from 03/24 where the new kernel version was introduced.

**Reply by AlvinSchiller** [reply]
> Yes, I think that's one option.
> 
> The other option could be that V2 is only supported for Bullseye anymore and we put more efforts in making future3 feature complete.

But we should observe the end of life for bullseye, and if v3 is a stable successor for v2 by then.
Some support changes to v2 could still be made later on, if neccessary.

**Reply by s-martin** [reply]
> > Yes, I think that's one option.
> > 
> > The other option could be that V2 is only supported for Bullseye anymore and we put more efforts in making future3 feature complete.
> 
> But we should observe the end of life for bullseye, and if v3 is a stable successor for v2 by then.

July is not so far away: https://wiki.debian.org/DebianReleases#Production_Releases 

> Some support changes to v2 could still be made later on, if neccessary.

👍

**Reply by AlvinSchiller** [reply]
At least it seems it goes into LTS security maintenance until 2026: https://wiki.debian.org/LTS

Just as a note: Buster is running out by June 30th, 2024

**Reply by varac** [reply]
I created a dedicated issue to focus on the GPIO issue: #2313

**Comment by hoffie** [comment]
Update regarding GPIO: The rpi-lpgio shim seems to have been added to the official repositories: https://github.com/raspberrypi/linux/issues/6037#issuecomment-2045225312

It's also worth reading the details from the shim's author down below the thread: https://github.com/raspberrypi/linux/issues/6037#issuecomment-2047494918

**Comment by UmePi** [comment]
Hi everyone,

First of all, thanks a lot for this great project!

Important disclaimer: I have some knowledge in coding but I'm really far from being an expert, so I'm sorry if the answers to my questions might seem obvious to some of you!

I have just started putting together my first Phoniebox but I'm running into a problem regarding the bluetooth function. I figured they might be related to the fact I'm using the **future3/main branch** on a **Raspberry Pi Model 3B with Bookworm Lite - 32bit, kernel 6.12.**

If I need to start a new discussion because this is not the right place, please just let me know :)

So basically, I can't seem to connect a bluetooth device (headset). I found a post where it was suggested to run "power on". But this is what I get:
```
$ bluetoothctl
[bluetooth]# power on
Failed to set power on: org.bluez.Error.Failed
```

On the bright side, the RFID reader is working, as I was able to add a card (however the songs could not be played via the browser interface - but I'll look into this issue later :D).

Has anyone faced a similar issue? Any suggestions about a fix or workaround for this? Or is downgrading to Bullseye the only reliable option? If so, how should I proceed?

Thanks a lot! :)

**Reply by AlvinSchiller** [reply]
> Hi everyone,
> 
> First of all, thanks a lot for this great project!
> 
> Important disclaimer: I have some knowledge in coding but I'm really far from being an expert, so I'm sorry if the answers to my questions might seem obvious to some of you!
> 
> I have just started putting together my first Phoniebox but I'm running into a problem regarding the bluetooth function. I figured they might be related to the fact I'm using the **future3/main branch** on a **Raspberry Pi Model 3B with Bookworm Lite - 32bit, kernel 6.12.**
> 
> If I need to start a new discussion because this is not the right place, please just let me know :)
> 
> So basically, I can't seem to connect a bluetooth device (headset). I found a post where it was suggested to run "power on". But this is what I get:
> 
> ```
> $ bluetoothctl
> [bluetooth]# power on
> Failed to set power on: org.bluez.Error.Failed
> ```
> 
> On the bright side, the RFID reader is working, as I was able to add a card (however the songs could not be played via the browser interface - but I'll look into this issue later :D).
> 
> Has anyone faced a similar issue? Any suggestions about a fix or workaround for this? Or is downgrading to Bullseye the only reliable option? If so, how should I proceed?
> 
> Thanks a lot! :)

Hi, this is probably a different topic and worth creating a new ticket.
Nevertheless have you tried the instructions from [this](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/main/documentation/builders/audio.md#bluetooth) guide and the notes in [this](https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Connecting_Bluetooth_device_to_Phoniebox) wiki post?

**Reply by UmePi** [reply]
Thanks! I opened a new ticket :) (I tried the instructions, but unfortunately it didn't work)
